File: Python/DynamicArray/dynamic_array_test.py
import unittest
from dynamic_array import DynamicArray

class TestDynamicArray(unittest.TestCase):

  # test add to back method
  def testAddToBack(self):
    case_1 = self.test_1.addToBack(12).arr 
    res_1 = [2,4,6,8,10,12]

    case_2 = self.test_2.addToBack(10).arr 
    res_2 = [1,2,3,4,5,6,7,8,9,10]

    case_3 = self.test_3.addToBack('pear').arr 
    res_3 = ['apple','banana','mango','grape','orange', 'pear']

    case_4 = self.test_4.addToBack({'name': 'Sarah', 'age': 24}).arr 
    res_4 = [{'name': 'John','age': 30},{'name': 'Janet','age': 40},{'name': 'Bob','age': 25},{'name': 'Sarah','age': 24}]

    case_5 = self.test_5.addToBack(True).arr 
    res_5 = [True, False, 'True', 'False', 0, 1, True]

    case_6 = self.test_6.addToBack(12).arr
    res_6 = [12]
    
    self.assertEqual(case_1, res_1)
    self.assertEqual(case_2, res_2)
    self.assertEqual(case_3, res_3)
    self.assertEqual(case_4, res_4)
    self.assertEqual(case_5, res_5)
    self.assertEqual(case_6, res_6)

  # test add to front method
  def testAddToFront(self):
    case_1 = self.test_1.addToFront(0).arr 
    res_1 = [0,2,4,6,8,10]

    case_2 = self.test_2.addToFront(0).arr 
    res_2 = [0,1,2,3,4,5,6,7,8,9]

    case_3 = self.test_3.addToFront('guava').arr 
    res_3 = ['guava','apple','banana','mango','grape','orange']

    case_4 = self.test_4.addToFront({'name': 'Jerry', 'age': 50}).arr 
    res_4 = [{'name': 'Jerry', 'age': 50},{'name': 'John','age': 30},{'name': 'Janet','age': 40},{'name': 'Bob','age': 25}]

    case_5 = self.test_5.addToFront(False).arr 
    res_5 = [False, True, False, 'True', 'False', 0, 1]

    case_6 = self.test_6.addToFront(24).arr
    res_6 = [24]
    
    self.assertEqual(case_1, res_1)
    self.assertEqual(case_2, res_2)
    self.assertEqual(case_3, res_3)
    self.assertEqual(case_4, res_4)
    self.assertEqual(case_5, res_5)
    self.assertEqual(case_6, res_6)

  # test add to remove back method
  def testRemoveBack(self):
    case_1 = self.test_1.removeBack().arr 
    res_1 = [2,4,6,8]

    case_2 = self.test_2.removeBack().arr 
    res_2 = [1,2,3,4,5,6,7,8]

    case_3 = self.test_3.removeBack().arr 
    res_3 = ['apple','banana','mango','grape']

    case_4 = self.test_4.removeBack().arr 
    res_4 = [{'name': 'John','age': 30},{'name': 'Janet','age': 40}]

    case_5 = self.test_5.removeBack().arr 
    res_5 = [True, False, 'True', 'False', 0]

    case_6 = self.test_6.removeBack().arr
    res_6 = []
    
    self.assertEqual(case_1, res_1)
    self.assertEqual(case_2, res_2)
    self.assertEqual(case_3, res_3)
    self.assertEqual(case_4, res_4)
    self.assertEqual(case_5, res_5)
    self.assertEqual(case_6, res_6)

  # test add to remove front method
  def testRemoveFront(self):
    case_1 = self.test_1.removeFront().arr 
    res_1 = [4,6,8,10]

    case_2 = self.test_2.removeFront().arr 
    res_2 = [2,3,4,5,6,7,8,9]

    case_3 = self.test_3.removeFront().arr 
    res_3 = ['banana','mango','grape','orange']

    case_4 = self.test_4.removeFront().arr 
    res_4 = [{'name': 'Janet','age': 40},{'name': 'Bob','age': 25}]

    case_5 = self.test_5.removeFront().arr 
    res_5 = [False, 'True', 'False', 0, 1]

    case_6 = self.test_6.removeFront().arr
    res_6 = []
    
    self.assertEqual(case_1, res_1)
    self.assertEqual(case_2, res_2)
    self.assertEqual(case_3, res_3)
    self.assertEqual(case_4, res_4)
    self.assertEqual(case_5, res_5)
    self.assertEqual(case_6, res_6)

  # test insert value method
  def testInsertValue(self):
    case_1 = self.test_1.insertValue(100, 3).arr 
    res_1 = [2,4,6,100,8,10]

    case_2 = self.test_2.insertValue(555, 5).arr 
    res_2 = [1,2,3,4,5,555,6,7,8,9]

    case_3 = self.test_3.insertValue('pineapple', 4).arr 
    res_3 = ['apple','banana','mango','grape','pineapple','orange']

    case_4 = self.test_4.insertValue({'name': 'Barb','age':60}, 2).arr 
    res_4 = [{'name': 'John','age': 30},{'name': 'Janet','age': 40},{'name': 'Barb','age': 60},{'name': 'Bob','age': 25}]

    case_5 = self.test_5.insertValue(False, 4).arr 
    res_5 = [True, False, 'True', 'False', False, 0, 1]

    case_6 = self.test_6.insertValue().arr
    res_6 = []
    
    self.assertEqual(case_1, res_1)
    self.assertEqual(case_2, res_2)
    self.assertEqual(case_3, res_3)
    self.assertEqual(case_4, res_4)
    self.assertEqual(case_5, res_5)
    self.assertEqual(case_6, res_6)

  # test remove index method
  def testRemoveIndex(self):
    case_1 = self.test_1.removeIndex(3).arr 
    res_1 = [2,4,6,10]

    case_2 = self.test_2.removeIndex(5).arr 
    res_2 = [1,2,3,4,5,7,8,9]

    case_3 = self.test_3.removeIndex(4).arr 
    res_3 = ['apple','banana','mango','grape']

    case_4 = self.test_4.removeIndex(1).arr 
    res_4 = [{'name': 'John','age': 30},{'name': 'Bob','age': 25}]

    case_5 = self.test_5.removeIndex(10).arr 
    res_5 = [True, False, 'True', 'False', 0, 1]

    case_6 = self.test_6.removeIndex(100).arr
    res_6 = []
    
    self.assertEqual(case_1, res_1)
    self.assertEqual(case_2, res_2)
    self.assertEqual(case_3, res_3)
    self.assertEqual(case_4, res_4)
    self.assertEqual(case_5, res_5)
    self.assertEqual(case_6, res_6)

  # test remove value method
  def testRemoveValue(self):
    case_1 = self.test_1.removeValue(8).arr 
    res_1 = [2,4,6,10]

    case_2 = self.test_2.removeValue(4).arr 
    res_2 = [1,2,3,5,6,7,8,9]

    case_3 = self.test_3.removeValue('mango').arr 
    res_3 = ['apple','banana','grape','orange']

    case_4 = self.test_4.removeValue({'name': 'Janet', 'age': 40}).arr 
    res_4 = [{'name': 'John','age': 30},{'name': 'Bob','age': 25}]

    case_5 = self.test_5.removeValue('False').arr 
    res_5 = [True, False, 'True',0, 1]

    case_6 = self.test_6.removeValue().arr
    res_6 = []
    
    self.assertEqual(case_1, res_1)
    self.assertEqual(case_2, res_2)
    self.assertEqual(case_3, res_3)
    self.assertEqual(case_4, res_4)
    self.assertEqual(case_5, res_5)
    self.assertEqual(case_6, res_6)

  # any task you want run before any method above is executed, put them in the setUp method
  def setUp(self):
    # add the setUp tasks
    print("running setUp")
    # set values
    self.test_1 = DynamicArray([2,4,6,8,10])
    self.test_2 = DynamicArray([1,2,3,4,5,6,7,8,9])
    self.test_3 = DynamicArray(['apple','banana','mango','grape','orange'])
    self.test_4 = DynamicArray([{'name': 'John','age': 30},{'name': 'Janet','age': 40},{'name': 'Bob','age': 25},])
    self.test_5 = DynamicArray([True, False, 'True', 'False', 0, 1])
    self.test_6 = DynamicArray([])
  # any task you want run after the tests are executed, put them in the tearDown method
  def tearDown(self):
    # add the tearDown tasks
    print("running tearDown tasks")
    # reset values
    self.test_1 = DynamicArray([2,4,6,8,10])
    self.test_2 = DynamicArray([1,2,3,4,5,6,7,8,9])
    self.test_3 = DynamicArray(['apple','banana','mango','grape','orange'])
    self.test_4 = DynamicArray([{'name': 'John','age': 30},{'name': 'Janet','age': 40},{'name': 'Bob','age': 25},])
    self.test_5 = DynamicArray([True, False, 'True', 'False', 0, 1])
    self.test_6 = DynamicArray([])
if __name__ == '__main__':
    unittest.main() # this runs our tests

File: JavaScript/DynamicArray/DynamicArray.test.js
// testing dynamic array with jest

const DynamicArray = require('./DynamicArray');


// arrays for testing
const testArrayOne = new DynamicArray();
const testArrayTwo = new DynamicArray();
const testArrayThree = new DynamicArray();
const testArrayFour = new DynamicArray();
const testArrayFive = new DynamicArray();
const testArraySix = new DynamicArray();

// reset arr for each test
const setVariables = _=> {
  testArrayOne.arr = [1, 2, 3, 4, 5, 6];
  testArrayTwo.arr = [2, 4, 6, 8, 10, 12, 14];
  testArrayThree.arr = ['apple', 'banana', 'mango', 'grape', 'orange'];
  testArrayFour.arr = [{ 'name': 'John', 'age': 30 }, { 'name': 'Janet', 'age': 40 }, { 'name': 'Bob', 'age': 25 }];
  testArrayFive.arr = [true, false, 'True', 'False', 0, 1];
  testArraySix.arr = [];
}

// set up
beforeEach(() => {
  setVariables();
});
// teardown
afterEach(() => {
  setVariables();
});

// tests for add to Front Method
const addToFrontTests = _=> {
  // test #1
  test('Test Add To Front Method #1', () => {
    const result = testArrayOne.addToFront(0);
    expect(result.arr).toEqual([0, 1, 2, 3, 4, 5, 6]);
  });

  //test #2
  test('Test Add To Front Method #2', () => {
    const result = testArrayTwo.addToFront(0);
    expect(result.arr).toEqual([0,2,4,6,8,10,12,14]);
  });

  //test #3
  test('Test Add To Front Method #3', () => {
    const result = testArrayThree.addToFront('guava');
    expect(result.arr).toEqual(['guava','apple', 'banana', 'mango', 'grape', 'orange']);
  });

  //test #4
  test('Test Add To Front Method #4', () => {
    const result = testArrayFour.addToFront({ 'name': 'Jerry', 'age': 50 });
    expect(result.arr).toEqual([{ 'name': 'Jerry', 'age': 50 }, { 'name': 'John', 'age': 30 }, { 'name': 'Janet', 'age': 40 }, { 'name': 'Bob', 'age': 25 }]);
  });

  //test #5
  test('Test Add To Front Method #5', () => {
    const result = testArrayFive.addToFront(false);
    expect(result.arr).toEqual([false,true, false, 'True', 'False', 0, 1]);
  });

  //test #6
  test('Test Add To Front Method #6', () => {
    const result = testArraySix.addToFront(100);
    expect(result.arr).toEqual([100]);
  });
}

// tests for add to Back Method
const addToBackTests = _=> {
  // test #1
  test('Test add To Back Method #1', () => {
    const result = testArrayOne.addToBack(7);
    expect(result.arr).toEqual([1, 2, 3, 4, 5, 6, 7]);
  });

  //test #2
  test('Test add To Back Method #2', () => {
    const result = testArrayTwo.addToBack(16);
    expect(result.arr).toEqual([2,4,6,8,10,12,14,16]);
  });

  //test #3
  test('Test add To Back Method #3', () => {
    const result = testArrayThree.addToBack('pear');
    expect(result.arr).toEqual(['apple', 'banana', 'mango', 'grape', 'orange', 'pear']);
  });

  //test #4
  test('Test add To Back Method #4', () => {
    const result = testArrayFour.addToBack({ 'name': 'Sarah', 'age': 24 });
    expect(result.arr).toEqual([{ 'name': 'John', 'age': 30 }, { 'name': 'Janet', 'age': 40 }, { 'name': 'Bob', 'age': 25 }, { 'name': 'Sarah', 'age': 24 }]);
  });

  //test #5
  test('Test add To Back Method #5', () => {
    const result = testArrayFive.addToBack(true);
    expect(result.arr).toEqual([true, false, 'True', 'False', 0, 1, true]);
  });

  //test #6
  test('Test add To Back Method #6', () => {
    const result = testArraySix.addToBack(100);
    expect(result.arr).toEqual([100]);
  });
}

// tests for remove Back Method
const removeBackTests = _=> {
  // test #1
  test('Test Remove Back Method #1', () => {
    const result = testArrayOne.removeBack();
    expect(result.arr).toEqual([1, 2, 3, 4, 5]);
  });

  //test #2
  test('Test Remove Back Method #2', () => {
    const result = testArrayTwo.removeBack();
    expect(result.arr).toEqual([2,4,6,8,10,12]);
  });

  //test #3
  test('Test Remove Back Method #3', () => {
    const result = testArrayThree.removeBack();
    expect(result.arr).toEqual(['apple', 'banana', 'mango', 'grape']);
  });

  //test #4
  test('Test Remove Back Method #4', () => {
    const result = testArrayFour.removeBack();
    expect(result.arr).toEqual([{ 'name': 'John', 'age': 30 }, { 'name': 'Janet', 'age': 40 }]);
  });

  //test #5
  test('Test Remove Back Method #5', () => {
    const result = testArrayFive.removeBack();
    expect(result.arr).toEqual([true, false, 'True', 'False', 0]);
  });

  //test #6
  test('Test Remove Back Method #6', () => {
    const result = testArraySix.removeBack();
    expect(result.arr).toEqual([]);
  });
}

// tests for remove Front Method
const removeFrontTests = _=> {
  // test #1
  test('Test Remove Front Method #1', () => {
    const result = testArrayOne.removeFront();
    expect(result.arr).toEqual([2, 3, 4, 5, 6]);
  });

  //test #2
  test('Test Remove Front Method #2', () => {
    const result = testArrayTwo.removeFront();
    expect(result.arr).toEqual([4,6,8,10,12,14]);
  });

  //test #3
  test('Test Remove Front Method #3', () => {
    const result = testArrayThree.removeFront();
    expect(result.arr).toEqual(['banana', 'mango', 'grape', 'orange']);
  });

  //test #4
  test('Test Remove Front Method #4', () => {
    const result = testArrayFour.removeFront();
    expect(result.arr).toEqual([{ 'name': 'Janet', 'age': 40 }, { 'name': 'Bob', 'age': 25 }]);
  });

  //test #5
  test('Test Remove Front Method #5', () => {
    const result = testArrayFive.removeFront();
    expect(result.arr).toEqual([false, 'True', 'False', 0, 1]);
  });

  //test #6
  test('Test Remove Front Method #6', () => {
    const result = testArraySix.removeFront();
    expect(result.arr).toEqual([]);
  });
}

// tests for Insert Method
const insertAtTests = _ => {
  // test #1
  test('Test Insert At Method #1', () => {
    const result = testArrayOne.insertAt(100, 4);
    expect(result.arr).toEqual([1, 2, 3, 4, 100, 5, 6]);
  });

  //test #2
  test('Test Insert At Method #2', () => {
    const result = testArrayTwo.insertAt('3', 1);
    expect(result.arr).toEqual([2, "3" ,4, 6, 8, 10, 12, 14]);
  });

  //test #3
  test('Test Insert At Method #3', () => {
    const result = testArrayThree.insertAt('grapefruit');
    expect(result.arr).toEqual(['apple', 'banana', 'mango', 'grape', 'orange']);
  });

  //test #4
  test('Test Insert At Method #4', () => {
    const result = testArrayFour.insertAt({ 'name': 'Sarah', 'age': 24 }, 1);
    expect(result.arr).toEqual([{ 'name': 'John', 'age': 30 }, { 'name': 'Sarah', 'age': 24 }, { 'name': 'Janet', 'age': 40 }, { 'name': 'Bob', 'age': 25 }]);
  });

  //test #5
  test('Test Insert At Method #5', () => {
    const result = testArrayFive.insertAt(true, 10);
    expect(result.arr).toEqual([true, false, 'True', 'False', 0, 1]);
  });

  //test #6
  test('Test Insert At Method #6', () => {
    const result = testArraySix.insertAt(100, 2);
    expect(result.arr).toEqual([]);
  });
}

// tests for Remove Index Method
const removeIndexTests = _ => {
  // test #1
  test('Test Remove Index Method #1', () => {
    const result = testArrayOne.removeIndex(4);
    expect(result.arr).toEqual([1, 2, 3, 4, 6]);
  });

  //test #2
  test('Test Remove Index Method #2', () => {
    const result = testArrayTwo.removeIndex(1);
    expect(result.arr).toEqual([2, 6, 8, 10, 12, 14]);
  });

  //test #3
  test('Test Remove Index Method #3', () => {
    const result = testArrayThree.removeIndex(2);
    expect(result.arr).toEqual(['apple', 'banana', 'grape', 'orange']);
  });

  //test #4
  test('Test Remove Index Method #4', () => {
    const result = testArrayFour.removeIndex(-1);
    expect(result.arr).toEqual([{ 'name': 'John', 'age': 30 }, { 'name': 'Janet', 'age': 40 }, { 'name': 'Bob', 'age': 25 }]);
  });

  //test #5
  test('Test Remove Index Method #5', () => {
    const result = testArrayFive.removeIndex(100);
    expect(result.arr).toEqual([true, false, 'True', 'False', 0, 1]);
  });

  //test #6
  test('Test Remove Index Method #6', () => {
    const result = testArraySix.removeIndex(10);
    expect(result.arr).toEqual([]);
  });
}

// run tests
addToFrontTests();
addToBackTests();
removeBackTests();
removeFrontTests();
insertAtTests();
removeIndexTests();